lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,342 @@
+      component->width = type.getBitWidthOrSentinel();
+      component->depth++;
+      break;
+    case Component::Type::Ground:
+      component->width = type.getBitWidthOrSentinel();
+      break;
+    }
+  }
+
+  if (unprocessedAnnos.empty())
+    return ArrayAttr::get(context, {});
+  else
+    return ArrayAttr::get(context, unprocessedAnnos);
+};
+
+/// Combined logic to handle Wires, Registers, and Nodes because these all use
+/// the same approach.
+void GrandCentralVisitor::handleRef(mlir::Operation *op) {
+  op->setAttr("annotations",
+              handleRefLike(op, AnnotationSet(op),
+                            op->getResult(0).getType().cast<FIRRTLType>()));
+};
+
+/// Remove Grand Central Annotations from ports of modules or external modules.
+/// Return argument attributes with annotations removed.
+ArrayAttr GrandCentralVisitor::handlePorts(Operation *op) {
+
+  SmallVector<Attribute> newArgAttrs;
+  auto ports = getModulePortInfo(op);
+  for (size_t i = 0, e = ports.size(); i != e; ++i) {
+    auto port = ports[i];
+    auto argAttr = mlir::function_like_impl::getArgAttrs(op, i);
+    if (!port.annotations) {
+      newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), argAttr));
+      continue;
+    }
+
+    auto remainingAnnotations = handleRefLike(
+        op, AnnotationSet(port.annotations.getValue()), port.type);
+
+    // Overwrite the annotations value with remaining annotations.
+    SmallVector<NamedAttribute> newArgAttr;
+    for (auto argIter = argAttr.begin(), argEnd = argAttr.end();
+         argIter != argEnd; ++argIter) {
+      if (argIter->first != "firrtl.annotations") {
+        newArgAttr.push_back(*argIter);
+        continue;
+      }
+      newArgAttr.push_back({argIter->first, remainingAnnotations});
+    }
+    newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), newArgAttr));
+  }
+
+  return ArrayAttr::get(op->getContext(), newArgAttrs);
+};
+
+void GrandCentralPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  AnnotationSet annotations(circuitOp);
+  if (annotations.empty())
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  // Store a mapping of interface name to InterfaceOp.
+  llvm::StringMap<sv::InterfaceOp> interfaces;
+
+  // Track the order that interfaces should be emitted in.
+  llvm::SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+
+  // Examine the Circuit's Annotations doing work to remove Grand Central
+  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
+  // Annotations with these when done.
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+  for (auto anno : annotations) {
+    auto clazz = anno.getClass();
+
+    if (clazz == "sifive.enterprise.grandcentral.AugmentedBundleType") {
+      auto defName = anno.getMember<StringAttr>("defName");
+      for (auto elt : anno.getMember<ArrayAttr>("elements")) {
```

Comment:
I don't know how to handle this. Everything about these annotations are checked during parsing/scattering. However, they are then assumed to be correct after that. E.g., the requirement that `"elements"` exists is [here](https://github.com/llvm/circt/blob/ea5dd62eae9ad303c936ad8be3cb9340f93f7ba0/lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp#L426).

I'd prefer to not unnecessarily duplicate checks in both places. However, annotations don't have verifiers so there's no guarantee that the annotations are in a valid state at this point.

I would prefer to move towards adding actual verifier hooks on annotations than duplicating checks in both places.

What do you think? 

Is this fine as is? Do annotation verifiers make sense? Should checks be duplicated until annotation verifiers exist?

Reply:
You're killing me.  There is a trivial way to do this with Swift's coding system, but doing this in C++ is a PITA.  This is analogous to turning stringly typed json etc into validated structured input.

I'd recommend emulating the Swift thing in C++, which is basically:

1) Define a C++ `AugmentedBundleType` (or whatever the right name is) struct that corresponds to the checked and unpacked annotation.
2) Give it a `static Optional<AugmentedBundleType> decode(Annotation annot)` member that does all the checking and decoding in one place, that returns nil on failure.
3) Make the caller check this.

This gets the grungy "walking through the annotation fields" logic out into one place,  making the logic for working with it more simple.

Reply:
> You're killing me.

I'm killing myself writing this stringly-typed soup. Just give me a product type pattern match with destructing please!

Solution seems clean :+1:.

The next level would be to move to actual `AugmentedBundleTypeAttr` that is checked and created by the parser instead of `DictionaryAttr` for the annotations. I know we discussed this and opted for the latter. However, maybe it's time to push in that direction?

Reply:
Updated to use structs and decoders that maybe generate the structs.